File: Project_and_Analysis_of_Algorithms/graph_search_applications/python/bfs.py
"""Breadth First Search"""

from collections import deque
from graphs import G1

def init_visited(graph: list[str, list[str]]) -> dict[str, bool]:
    """Maps all nodes as unvisited."""
    return {node: False for node in graph.keys()}


def bfs(graph: list[str, list[str]], start_node: str, visited: dict[str, bool]):
    """Iterative Breadth First Search starting from the start node using a queue."""
    queue = deque(start_node)
    visited[start_node] = True

    while len(queue) > 0:
        cur_node: str = queue.popleft()
        print(cur_node) # Process
        for neighbour in graph[cur_node]:
            if not visited[neighbour]:
                queue.append(neighbour)
                visited[neighbour] = True


def main():
    bfs(graph=G1, start_node="A", visited=init_visited(G1))


if __name__ == "__main__":
    main()

File: Project_and_Analysis_of_Algorithms/graph_search_applications/c++/bfs.cpp
#include "graphs.h"

// initVisited maps all nodes as unvisited.
visitedMap(string) initVisited(adjacencyMap(string) graph) {
    visitedMap(string) visited;
    for (auto const&[node, neighbous] : graph) {
        visited[node] = false;
    }
    return visited;
}

// Iterative breadth first search from the starting node using a queue.
void bfs(adjacencyMap(string) graph, string startNode, visitedMap(string) visited) {
    queue<string> q;
    q.push(startNode);
    visited[startNode] = true;

    while(!q.empty()) {
        string curNode = q.front(); q.pop();
        cout << curNode << "\n"; // Process Node.
        for (string neighbour : graph[curNode]) {
            if (!visited[neighbour]){
                q.push(neighbour);
                visited[neighbour] = true;
            }
        }
    }
}

int main(void) {
    bfs(G1, "A", initVisited(G1));
    return EXIT_SUCCESS;
}
